File: cpp/src/common/postprocess.cpp
#include "edgeai/common/postprocess.hpp"

#include "edgeai/common/preprocess.hpp"

#include <algorithm>
#include <cmath>
#include <numeric>
#include <stdexcept>
#include <utility>

namespace edgeai::common {
namespace {

struct Candidate {
    std::size_t candidate_index{0};
    int class_id{-1};
    float objectness{0.0F};
    float class_score{0.0F};
    float confidence{0.0F};
    Box xywh;
    Box xyxy;
};

float round_six(float value) {
    return static_cast<float>(std::round(static_cast<double>(value) * 1'000'000.0) /
                              1'000'000.0);
}

bool finite_box(const Box& box) {
    return std::isfinite(box.x1) && std::isfinite(box.y1) && std::isfinite(box.x2) &&
           std::isfinite(box.y2);
}

}  // namespace

Box xywh_to_xyxy(const Box& box) {
    if (!finite_box(box)) {
        throw std::runtime_error("xywh box contains a non-finite value");
    }
    return {
        box.x1 - box.x2 / 2.0F,
        box.y1 - box.y2 / 2.0F,
        box.x1 + box.x2 / 2.0F,
        box.y1 + box.y2 / 2.0F,
    };
}

float box_iou(const Box& left, const Box& right) {
    if (!finite_box(left) || !finite_box(right)) {
        throw std::runtime_error("IoU box contains a non-finite value");
    }
    const float intersection_width =
        std::max(0.0F, std::min(left.x2, right.x2) - std::max(left.x1, right.x1));
    const float intersection_height =
        std::max(0.0F, std::min(left.y2, right.y2) - std::max(left.y1, right.y1));
    const float intersection = intersection_width * intersection_height;
    const float left_area =
        std::max(0.0F, left.x2 - left.x1) * std::max(0.0F, left.y2 - left.y1);
    const float right_area =
        std::max(0.0F, right.x2 - right.x1) * std::max(0.0F, right.y2 - right.y1);
    const float union_area = left_area + right_area - intersection;
    return union_area > 0.0F ? intersection / union_area : 0.0F;
}

PostprocessResult decode_yolov5_output(
    const std::vector<float>& raw_output,
    const std::vector<std::int64_t>& output_shape,
    const std::vector<std::string>& class_names,
    const LetterboxMetadata& metadata,
    const InferenceConfig& config
) {
    validate_config(config);
    if (output_shape.size() != 3U || output_shape[0] != 1 || output_shape[1] < 0 ||
        output_shape[2] != static_cast<std::int64_t>(5U + class_names.size())) {
        throw std::runtime_error("YOLOv5 output shape is incompatible with batch-1 classes");
    }
    const auto candidate_count = static_cast<std::size_t>(output_shape[1]);
    const auto attribute_count = static_cast<std::size_t>(output_shape[2]);
    if (raw_output.size() != candidate_count * attribute_count) {
        throw std::runtime_error("YOLOv5 output buffer size does not match its shape");
    }
    if (class_names.empty()) {
        throw std::runtime_error("YOLOv5 class names must not be empty");
    }

    std::vector<Candidate> candidates;
    candidates.reserve(candidate_count);
    for (std::size_t candidate_index = 0; candidate_index < candidate_count; ++candidate_index) {
        const std::size_t offset = candidate_index * attribute_count;
        for (std::size_t attribute = 0; attribute < attribute_count; ++attribute) {
            if (!std::isfinite(raw_output[offset + attribute])) {
                throw std::runtime_error("YOLOv5 output contains a non-finite value");
            }
        }
        const float objectness = raw_output[offset + 4U];
        if (objectness < 0.0F || objectness > 1.0F) {
            throw std::runtime_error("YOLOv5 objectness falls outside [0, 1]");
        }
        int class_id = 0;
        float class_score = raw_output[offset + 5U];
        for (std::size_t class_index = 0; class_index < class_names.size(); ++class_index) {
            const float score = raw_output[offset + 5U + class_index];
            if (score < 0.0F || score > 1.0F) {
                throw std::runtime_error("YOLOv5 class score falls outside [0, 1]");
            }
            if (score > class_score) {
                class_score = score;
                class_id = static_cast<int>(class_index);
            }
        }
        const float confidence = objectness * class_score;
        if (static_cast<double>(confidence) < config.confidence_threshold) {
            continue;
        }
        const Box xywh{
            raw_output[offset],
            raw_output[offset + 1U],
            raw_output[offset + 2U],
            raw_output[offset + 3U],
        };
        if (xywh.x2 <= 0.0F || xywh.y2 <= 0.0F) {
            throw std::runtime_error("thresholded YOLOv5 candidate has non-positive box size");
        }
        candidates.push_back(
            {candidate_index,
             class_id,
             objectness,
             class_score,
             confidence,
             xywh,
             xywh_to_xyxy(xywh)}
        );
    }

    std::vector<std::size_t> order(candidates.size());
    std::iota(order.begin(), order.end(), 0U);
    std::sort(order.begin(), order.end(), [&](std::size_t left_index, std::size_t right_index) {
        const Candidate& left = candidates[left_index];
        const Candidate& right = candidates[right_index];
        if (left.confidence != right.confidence) {
            return left.confidence > right.confidence;
        }
        if (left.class_id != right.class_id) {
            return left.class_id < right.class_id;
        }
        return left.candidate_index < right.candidate_index;
    });

    std::vector<std::size_t> kept;
    kept.reserve(std::min(candidates.size(), static_cast<std::size_t>(config.max_detections)));
    for (const std::size_t position : order) {
        if (kept.size() >= static_cast<std::size_t>(config.max_detections)) {
            break;
        }
        bool suppressed = false;
        for (const std::size_t kept_position : kept) {
            if (candidates[kept_position].class_id == candidates[position].class_id &&
                static_cast<double>(box_iou(
                    candidates[position].xyxy,
                    candidates[kept_position].xyxy
                )) > config.iou_threshold) {
                suppressed = true;
                break;
            }
        }
        if (!suppressed) {
            kept.push_back(position);
        }
    }

    PostprocessResult result;
    result.raw_candidate_count = candidate_count;
    result.threshold_candidate_count = candidates.size();
    result.nms_candidate_count = kept.size();
    result.detections.reserve(kept.size());
    for (const std::size_t position : kept) {
        const Candidate& candidate = candidates[position];
        const auto restored = restore_and_clip_box(candidate.xyxy, metadata);
        if (!restored.has_value()) {
            ++result.invalid_box_count;
            continue;
        }
        Detection detection;
        detection.rank = result.detections.size() + 1U;
        detection.candidate_index = candidate.candidate_index;
        detection.class_id = candidate.class_id;
        detection.class_name = class_names[static_cast<std::size_t>(candidate.class_id)];
        detection.objectness = round_six(candidate.objectness);
        detection.class_score = round_six(candidate.class_score);
        detection.confidence = round_six(candidate.confidence);
        detection.box_xywh_input = {
            round_six(candidate.xywh.x1),
            round_six(candidate.xywh.y1),
            round_six(candidate.xywh.x2),
            round_six(candidate.xywh.y2),
        };
        detection.box_xyxy_input = {
            round_six(candidate.xyxy.x1),
            round_six(candidate.xyxy.y1),
            round_six(candidate.xyxy.x2),
            round_six(candidate.xyxy.y2),
        };
        detection.box_xyxy_source = {
            round_six(restored->x1),
            round_six(restored->y1),
            round_six(restored->x2),
            round_six(restored->y2),
        };
        result.detections.push_back(std::move(detection));
    }
    return result;
}

}  // namespace edgeai::common

File: python/edgeai_benchmark/postprocess.py
"""Backend-independent YOLOv5 detection postprocessing."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import numpy as np


class PostprocessError(RuntimeError):
    """Raised when raw output or postprocessing configuration is invalid."""


@dataclass(frozen=True)
class DetectionConfig:
    confidence_threshold: float
    iou_threshold: float
    class_aware_nms: bool
    max_detections: int

    def validate(self) -> None:
        if not 0.0 <= self.confidence_threshold <= 1.0:
            raise PostprocessError("confidence threshold must be in [0, 1]")
        if not 0.0 <= self.iou_threshold <= 1.0:
            raise PostprocessError("IoU threshold must be in [0, 1]")
        if self.class_aware_nms is not True:
            raise PostprocessError("Task 004 requires class-aware NMS")
        if self.max_detections <= 0:
            raise PostprocessError("maximum detections must be positive")


def xywh_to_xyxy(boxes: np.ndarray) -> np.ndarray:
    """Convert center-width-height boxes into corner boxes."""
    if boxes.ndim != 2 or boxes.shape[1] != 4:
        raise PostprocessError(f"expected Nx4 xywh boxes, observed {boxes.shape}")
    converted = np.empty_like(boxes, dtype=np.float32)
    converted[:, 0] = boxes[:, 0] - boxes[:, 2] / 2.0
    converted[:, 1] = boxes[:, 1] - boxes[:, 3] / 2.0
    converted[:, 2] = boxes[:, 0] + boxes[:, 2] / 2.0
    converted[:, 3] = boxes[:, 1] + boxes[:, 3] / 2.0
    return converted


def box_iou(box: np.ndarray, boxes: np.ndarray) -> np.ndarray:
    """Compute IoU between one xyxy box and an Nx4 array."""
    if box.shape != (4,) or boxes.ndim != 2 or boxes.shape[1] != 4:
        raise PostprocessError("IoU expects one xyxy box and an Nx4 box array")
    intersection_width = np.maximum(
        0.0,
        np.minimum(box[2], boxes[:, 2]) - np.maximum(box[0], boxes[:, 0]),
    )
    intersection_height = np.maximum(
        0.0,
        np.minimum(box[3], boxes[:, 3]) - np.maximum(box[1], boxes[:, 1]),
    )
    intersection = intersection_width * intersection_height
    box_area = max(0.0, float(box[2] - box[0])) * max(0.0, float(box[3] - box[1]))
    boxes_area = np.maximum(0.0, boxes[:, 2] - boxes[:, 0]) * np.maximum(
        0.0, boxes[:, 3] - boxes[:, 1]
    )
    union = box_area + boxes_area - intersection
    return np.divide(
        intersection,
        union,
        out=np.zeros_like(intersection, dtype=np.float32),
        where=union > 0.0,
    )


def class_aware_nms(
    boxes: np.ndarray,
    scores: np.ndarray,
    class_ids: np.ndarray,
    candidate_indices: np.ndarray,
    iou_threshold: float,
    max_detections: int,
) -> list[int]:
    """Return deterministic candidate positions ordered by descending confidence."""
    count = len(scores)
    if boxes.shape != (count, 4):
        raise PostprocessError("NMS boxes and scores have inconsistent lengths")
    if class_ids.shape != (count,) or candidate_indices.shape != (count,):
        raise PostprocessError("NMS metadata and scores have inconsistent lengths")
    order = np.lexsort((candidate_indices, class_ids, -scores))
    kept: list[int] = []
    for position in order.tolist():
        if len(kept) >= max_detections:
            break
        same_class_kept = [index for index in kept if class_ids[index] == class_ids[position]]
        if same_class_kept:
            overlaps = box_iou(boxes[position], boxes[same_class_kept])
            if bool(np.any(overlaps > iou_threshold)):
                continue
        kept.append(position)
    return kept


def restore_and_clip_box(
    box: np.ndarray,
    letterbox_metadata: dict[str, Any],
) -> np.ndarray | None:
    """Undo letterbox geometry and clip a box to source-image boundaries."""
    scale = float(letterbox_metadata["scale"])
    padding = letterbox_metadata["padding"]
    original_size = letterbox_metadata["original_size"]
    if scale <= 0.0:
        raise PostprocessError(f"invalid letterbox scale: {scale}")
    restored = np.asarray(box, dtype=np.float32).copy()
    restored[[0, 2]] = (restored[[0, 2]] - float(padding["left"])) / scale
    restored[[1, 3]] = (restored[[1, 3]] - float(padding["top"])) / scale
    width = float(original_size["width"])
    height = float(original_size["height"])
    restored[[0, 2]] = np.clip(restored[[0, 2]], 0.0, width)
    restored[[1, 3]] = np.clip(restored[[1, 3]], 0.0, height)
    if restored[2] <= restored[0] or restored[3] <= restored[1]:
        return None
    return restored


def _rounded(value: float) -> float:
    return round(float(value), 6)


def decode_yolov5_output(
    raw_output: np.ndarray,
    class_names: list[str],
    letterbox_metadata: dict[str, Any],
    config: DetectionConfig,
) -> dict[str, Any]:
    """Decode one verified YOLOv5 raw output into deterministic detections."""
    config.validate()
    if raw_output.ndim != 3 or raw_output.shape[0] != 1:
        raise PostprocessError(
            f"expected batch-1 rank-3 raw output, observed {raw_output.shape}"
        )
    expected_attributes = 5 + len(class_names)
    if raw_output.shape[2] != expected_attributes:
        raise PostprocessError(
            f"expected {expected_attributes} attributes, observed {raw_output.shape[2]}"
        )
    if raw_output.dtype != np.float32:
        raise PostprocessError(f"expected float32 raw output, observed {raw_output.dtype}")
    if not np.isfinite(raw_output).all():
        raise PostprocessError("raw output contains non-finite values")

    candidates = raw_output[0]
    objectness = candidates[:, 4]
    class_scores = candidates[:, 5:]
    if np.any(objectness < 0.0) or np.any(objectness > 1.0):
        raise PostprocessError("objectness values fall outside [0, 1]")
    if np.any(class_scores < 0.0) or np.any(class_scores > 1.0):
        raise PostprocessError("class scores fall outside [0, 1]")
    class_ids = np.argmax(class_scores, axis=1).astype(np.int64)
    selected_class_scores = class_scores[np.arange(len(candidates)), class_ids]
    confidences = objectness * selected_class_scores
    threshold_mask = confidences >= config.confidence_threshold
    candidate_indices = np.nonzero(threshold_mask)[0].astype(np.int64)

    if candidate_indices.size == 0:
        return {
            "raw_candidate_count": int(len(candidates)),
            "threshold_candidate_count": 0,
            "nms_candidate_count": 0,
            "invalid_box_count": 0,
            "detections": [],
        }

    selected_xywh = candidates[candidate_indices, :4]
    if np.any(selected_xywh[:, 2:] <= 0.0):
        raise PostprocessError("thresholded candidates contain non-positive box sizes")
    boxes = xywh_to_xyxy(selected_xywh)
    selected_confidences = confidences[candidate_indices]
    selected_class_ids = class_ids[candidate_indices]
    kept_positions = class_aware_nms(
        boxes,
        selected_confidences,
        selected_class_ids,
        candidate_indices,
        config.iou_threshold,
        config.max_detections,
    )

    detections = []
    invalid_box_count = 0
    for position in kept_positions:
        restored = restore_and_clip_box(boxes[position], letterbox_metadata)
        if restored is None:
            invalid_box_count += 1
            continue
        candidate_index = int(candidate_indices[position])
        class_id = int(selected_class_ids[position])
        input_box = boxes[position]
        input_xywh = selected_xywh[position]
        detections.append(
            {
                "candidate_index": candidate_index,
                "class_id": class_id,
                "class_name": class_names[class_id],
                "objectness": _rounded(objectness[candidate_index]),
                "class_score": _rounded(selected_class_scores[candidate_index]),
                "confidence": _rounded(selected_confidences[position]),
                "box_xywh_input": [_rounded(value) for value in input_xywh],
                "box_xyxy_input": [_rounded(value) for value in input_box],
                "box_xyxy_source": [_rounded(value) for value in restored],
            }
        )

    for rank, detection in enumerate(detections, start=1):
        detection["rank"] = rank
    return {
        "raw_candidate_count": int(len(candidates)),
        "threshold_candidate_count": int(len(candidate_indices)),
        "nms_candidate_count": int(len(kept_positions)),
        "invalid_box_count": invalid_box_count,
        "detections": detections,
    }
